File: java/Samples/Biometrics/VB/SimpleFingersSampleVB/GeneralizeFinger.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.IO
Imports System.Linq
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Images
Imports Neurotec.Biometrics.Gui

Partial Public Class GeneralizeFinger
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "Private fields"

	Private _subject As NSubject = Nothing

#End Region

#Region "Public properties"

	Private privateBiometricClient As NBiometricClient
	Public Property BiometricClient() As NBiometricClient
		Get
			Return privateBiometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			privateBiometricClient = value
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Sub OnCreateTemplateCompleted(ByVal result As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnCreateTemplateCompleted), result)
		Else
			Try
				Dim status As NBiometricStatus = BiometricClient.EndCreateTemplate(result)
				If status = NBiometricStatus.Ok Then
					lblStatus.Text = "Status: Ok"
					lblStatus.ForeColor = Color.Green
					btnSaveTemplate.Enabled = True
					fingerView.Finger = _subject.Fingers.Last()
				Else
					lblStatus.Text = String.Format("Status: {0}", status)
					lblStatus.ForeColor = Color.Red
				End If
			Catch ex As Exception
				lblStatus.Text = "Status: error occured"
				lblStatus.ForeColor = Color.Red
				Utils.ShowException(ex)
			End Try
			btnOpenImages.Enabled = True
		End If
	End Sub

	Private Sub BtnOpenImagesClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnOpenImages.Click
		_subject = Nothing
		fingerView.Finger = Nothing
		btnSaveTemplate.Enabled = False
		lblImageCount.Text = "0"
		lblStatus.Visible = False

		If openFileDialog.ShowDialog() = DialogResult.OK Then
			Dim files = openFileDialog.FileNames
			If files.Length < 3 OrElse files.Length > 10 Then
				Dim msg As String = String.Format("{0} images selected. Please select at least 3 and no more than 10 images", If(files.Length > 10, "Too many", "Too few"))
				MessageBox.Show(msg, Text, MessageBoxButtons.OK, MessageBoxIcon.Warning)
			Else
				_subject = New NSubject()
				For Each fileName In files
					Dim finger As New NFinger() With {.FileName = fileName, .SessionId = 1}
					_subject.Fingers.Add(finger)
				Next fileName
				lblImageCount.Text = openFileDialog.FileNames.Length.ToString()
				lblStatus.Text = "Status: performing extraction and generalizarion"
				lblStatus.Visible = True
				lblStatus.ForeColor = Color.Black
				BiometricClient.BeginCreateTemplate(_subject, AddressOf OnCreateTemplateCompleted, Nothing)
				btnOpenImages.Enabled = False
			End If
		End If
	End Sub

	Private Sub GeneralizeFingerLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		If (Not DesignMode) Then
			openFileDialog.Filter = NImages.GetOpenFileFilterString(True, True)
		End If
	End Sub

	Private Sub BtnSaveTemplateClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveTemplate.Click
		If _subject IsNot Nothing Then
			If saveFileDialog.ShowDialog() = DialogResult.OK Then
				Try
					Using templateBuffer = _subject.GetTemplateBuffer()
						File.WriteAllBytes(saveFileDialog.FileName, templateBuffer.ToArray())
					End Using
				Catch ex As Exception
					Utils.ShowException(ex)
				End Try
			End If
		End If
	End Sub

	Private Sub GeneralizeFingerVisibleChanged(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.VisibleChanged
		If Visible AndAlso BiometricClient IsNot Nothing Then
			BiometricClient.FingersReturnBinarizedImage = True
		End If
	End Sub

	Private Sub ChbShowBinarizedImageCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chbShowBinarizedImage.CheckedChanged
		fingerView.ShownImage = If(chbShowBinarizedImage.Checked, ShownImage.Result, ShownImage.Original)
	End Sub

#End Region
End Class

File: java/Samples/Biometrics/CS/SimpleFingersSampleCS/GeneralizeFinger.cs
using System;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Images;
using Neurotec.Biometrics.Gui;

namespace Neurotec.Samples
{
	public partial class GeneralizeFinger : UserControl
	{
		#region Public constructor

		public GeneralizeFinger()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NSubject _subject = null;

		#endregion

		#region Public properties

		public NBiometricClient BiometricClient { get; set; }

		#endregion

		#region Private methods

		private void OnCreateTemplateCompleted(IAsyncResult result)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnCreateTemplateCompleted), result);
			}
			else
			{
				try
				{
					NBiometricStatus status = BiometricClient.EndCreateTemplate(result);
					if (status == NBiometricStatus.Ok)
					{
						lblStatus.Text = "Status: Ok";
						lblStatus.ForeColor = Color.Green;
						btnSaveTemplate.Enabled = true;
						fingerView.Finger = _subject.Fingers.Last();
					}
					else
					{
						lblStatus.Text = string.Format("Status: {0}", status);
						lblStatus.ForeColor = Color.Red;
					}
				}
				catch (Exception ex)
				{
					lblStatus.Text = "Status: error occured";
					lblStatus.ForeColor = Color.Red;
					Utils.ShowException(ex);
				}
				btnOpenImages.Enabled = true;
			}
		}

		private void BtnOpenImagesClick(object sender, EventArgs e)
		{
			_subject = null;
			fingerView.Finger = null;
			btnSaveTemplate.Enabled = false;
			lblImageCount.Text = "0";
			lblStatus.Visible = false;

			if (openFileDialog.ShowDialog() == DialogResult.OK)
			{
				var files = openFileDialog.FileNames;
				if (files.Length < 3 || files.Length > 10)
				{
					string msg = string.Format("{0} images selected. Please select at least 3 and no more than 10 images", files.Length > 10 ? "Too many" : "Too few");
					MessageBox.Show(msg, Text, MessageBoxButtons.OK, MessageBoxIcon.Warning);
				}
				else
				{
					_subject = new NSubject();
					foreach (var name in files)
					{
						NFinger finger = new NFinger()
						{
							FileName = name,
							SessionId = 1 // all fingers with same session will be generalized
						};
						_subject.Fingers.Add(finger);
					}
					lblImageCount.Text = openFileDialog.FileNames.Length.ToString();
					lblStatus.Text = "Status: performing extraction and generalizarion";
					lblStatus.Visible = true;
					lblStatus.ForeColor = Color.Black;
					BiometricClient.BeginCreateTemplate(_subject, OnCreateTemplateCompleted, null);
					btnOpenImages.Enabled = false;
				}
			}
		}

		private void GeneralizeFingerLoad(object sender, EventArgs e)
		{
			if (!DesignMode)
			{
				openFileDialog.Filter = NImages.GetOpenFileFilterString(true, true);
			}
		}

		private void BtnSaveTemplateClick(object sender, EventArgs e)
		{
			if (_subject != null)
			{
				if (saveFileDialog.ShowDialog() == DialogResult.OK)
				{
					try
					{
						using (var templateBuffer = _subject.GetTemplateBuffer())
						{
							File.WriteAllBytes(saveFileDialog.FileName, templateBuffer.ToArray());
						}
					}
					catch (Exception ex)
					{
						Utils.ShowException(ex);
					}
				}
			}
		}

		private void GeneralizeFingerVisibleChanged(object sender, EventArgs e)
		{
			if (Visible && BiometricClient != null)
			{
				BiometricClient.FingersReturnBinarizedImage = true;
			}
		}

		private void ChbShowBinarizedImageCheckedChanged(object sender, EventArgs e)
		{
			fingerView.ShownImage = chbShowBinarizedImage.Checked ? ShownImage.Result : ShownImage.Original;
		}

		#endregion
	}
}
